elastic/kibana
# [Security solution] Fix streaming on cloud

## Summary

Unfortunately, the strategy to use a TransformStream in the `invokeStream` method (see [1](https://github.com/elastic/kibana/blob/main/x-pack/plugins/stack_connectors/server/connector_types/bedrock/bedrock.ts#L186), [2](https://github.com/elastic/kibana/blob/main/x-pack/plugins/stack_connectors/server/connector_types/openai/openai.ts#L207)) was unsuccessful in production. The transform caused additional latency and caused the stream not to work at all when passing through the cloud proxy. I removed the TransformStream from the stack_connectors, and instead we need to count on the consumer to properly parse the response stream.

### Bedrock

Bedrock will sometimes send multiple messages in one buffer. To account for this, I again followed the [langchain](https://github.com/langchain-ai/langchainjs/pull/3364) bedrock extension and used a concatenated buffer to process multiple messages one at a time.

### To test

1. Test with the [deployment](https://kibana-pr-171578.kb.us-west2.gcp.elastic-cloud.com:9243/app/security/get_started) on this PR
2. Use `vault` to access creds: `vault read secret/kibana-issues/dev/cloud-deploy/kibana-pr-171578`
3. Ensure Knowledge Base is turned off
4. Test with both OpenAI and Bedrock credentials, responses should stream
5. Test stop generating and regenerate buttons
6. I also created connectors with bad model. Test that you receive an error with the connectors labeled "bad model"
